Add offline schema updates for ClickHouse

bnaecker opened the pull request:
- Some cleanup around issuing multiple SQL statements from a file
- Create directory structure for storing schema updates modeled after CRDB up.sql files, but using integer versions, and move all existing SQL into version 2
- Add version 3, which fixes https://github.com/oxidecomputer/omicron/issues/4369, but does not apply it yet
- Add methods in the client for listing, reading, and applying one or more updates to the oximeter database from the upgrade files
- Add tests for upgrade application
- Add `clickhouse-schema-updater` binary for running them on demand
- Modify `oximeter-collector` to _not_ wipe / reinit the DB on startup if the version has change, but instead wait for the version to be equal to what it is compiled against. This relies on updates from the developer being applied before `oximeter` will continue.

karencfv commented:
Will this PR close https://github.com/oxidecomputer/omicron/issues/4048 as well?

bnaecker commented:
@karencfv Yep, I think so. I still need do a bunch of testing and verification of the actual binary, but I've got a good set of tests around the internal implementation of that in the `oximeter_db::Client` tests.

karencfv commented:
Nice!!! :tada:

bnaecker commented:
### Testing setup

In addition to the unit tests inside the actual database client, I also tested this several times on my local developer machine. I started with the control plane built and installed from `main` (73c3571fe591af14bbe9c341f6ed4a4169446f91). We can see here that ClickHouse is running against version 2:

```bash
bnaecker@shale : ~/omicron $ pfexec zlogin oxz_clickhouse_de7eb711-5b35-4b64-894c-7d3a4764c189
[Connected to zone 'oxz_clickhouse_de7eb711-5b35-4b64-894c-7d3a4764c189' pts/3]
The illumos Project     helios-2.0.22248        October 2023
root@oxz_clickhouse_de7eb711-5b35-4b64-894c-7d3a4764c189:~# ipadm
ADDROBJ           TYPE     STATE        ADDR
lo0/v4            static   ok           127.0.0.1/8
lo0/v6            static   ok           ::1/128
oxControlService19/ll addrconf ok       fe80::8:20ff:fe9e:c080%oxControlService19/10
oxControlService19/omicron6 static ok   fd00:1122:3344:101::e/64
root@oxz_clickhouse_de7eb711-5b35-4b64-894c-7d3a4764c189:~# /opt/oxide/clickhouse/clickhouse client --database oximeter --host fd00:1122:3344:101::e
ClickHouse client version 22.8.9.1.
Connecting to database oximeter at fd00:1122:3344:101::e:9000 as user default.
Connected to ClickHouse server version 22.8.9 revision 54460.

oxz_clickhouse_de7eb711-5b35-4b64-894c-7d3a4764c189 :) select * from version;

SELECT *
FROM version

Query id: 90d30c65-6927-41ec-a91e-7e9ba5e71c42

┌─value─┬─────────────────────timestamp─┐
│     2 │ 2023-10-31 19:03:05.000000000 │
└───────┴───────────────────────────────┘

1 row in set. Elapsed: 0.001 sec.

oxz_clickhouse_de7eb711-5b35-4b64-894c-7d3a4764c189 :)
```

`oximeter` is happily inserting records:

```bash
bnaecker@shale : ~/omicron $ pfexec zlogin oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606 'tail $(svcs -L oximeter)'
{"msg":"unrolling 131 total samples","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:09:15.754578608Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":4644,"id":"535d567e-f5d8-42b9-92bc-c24dfc88d607","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent"}
{"msg":"inserted rows into table","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:09:15.767781162Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":4644,"id":"535d567e-f5d8-42b9-92bc-c24dfc88d607","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","table_name":"oximeter.fields_i64","n_rows":40}
{"msg":"inserted rows into table","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:09:15.772416835Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":4644,"id":"535d567e-f5d8-42b9-92bc-c24dfc88d607","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","table_name":"oximeter.fields_string","n_rows":198}
{"msg":"inserted rows into table","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:09:15.776443983Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":4644,"id":"535d567e-f5d8-42b9-92bc-c24dfc88d607","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","table_name":"oximeter.fields_uuid","n_rows":325}
{"msg":"inserted rows into table","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:09:15.777773065Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":4644,"id":"535d567e-f5d8-42b9-92bc-c24dfc88d607","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","table_name":"oximeter.measurements_bool","n_rows":6}
{"msg":"inserted rows into table","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:09:15.779191948Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":4644,"id":"535d567e-f5d8-42b9-92bc-c24dfc88d607","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","table_name":"oximeter.measurements_cumulativei64","n_rows":70}
{"msg":"inserted rows into table","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:09:15.780728493Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":4644,"id":"535d567e-f5d8-42b9-92bc-c24dfc88d607","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","table_name":"oximeter.measurements_histogramf64","n_rows":34}
{"msg":"inserted rows into table","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:09:15.782027185Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":4644,"id":"535d567e-f5d8-42b9-92bc-c24dfc88d607","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","table_name":"oximeter.measurements_i64","n_rows":21}
{"msg":"collecting from producer","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:09:20.995449246Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":4644,"producer_id":"52028f5e-8f19-461a-8624-1b1450f26654","component":"collection-task","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent"}
{"msg":"collected 3 total results","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:09:20.996114992Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":4644,"producer_id":"52028f5e-8f19-461a-8624-1b1450f26654","component":"collection-task","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent"}
```

### Installing new code

To test the new upgrade code, I then rebuilt `oximeter` and the `clickhouse-schema-updater` binaries from this PR commit (caed2ee1abbe41e39ae1cddab5a282083f9392f9); stopped the `oximeter` service; and copied the new binaries and schema directories into the zone.

One of the new features of `oximeter` is that on startup it waits until the DB matches the version it was compiled against. This is similar to the current implementation in Nexus. So when I restarted `oximeter`, we see this:

```bash
[ Oct 31 19:17:36 Enabled. ]
[ Oct 31 19:17:36 Executing start method ("ctrun -l child -o noorphan,regent /opt/oxide/oximeter/bin/oximeter run /var/svc/manifest/site/oximeter/config.toml --address [fd00:1122:3344:101::d]:12223 --id f3a5ade4-a165-4ecd-a365-e57b3cdf3606 &"). ]
[ Oct 31 19:17:36 Method "start" exited with status 0. ]
note: configured to log to "/dev/stdout"
{"msg":"registered DTrace probes","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:17:36.165385878Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288}
{"msg":"starting oximeter server","v":0,"name":"oximeter","level":30,"time":"2023-10-31T19:17:36.165806662Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"file":"oximeter/collector/src/lib.rs:760"}
{"msg":"new DNS resolver","v":0,"name":"oximeter","level":30,"time":"2023-10-31T19:17:36.165924353Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"component":"DnsResolver","file":"internal-dns/src/resolver.rs:60","addresses":"[[fd00:1122:3344:1::1]:53, [fd00:1122:3344:2::1]:53, [fd00:1122:3344:3::1]:53, [fd00:1122:3344:4::1]:53, [fd00:1122:3344:5::1]:53]"}
{"msg":"creating ClickHouse client","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:17:36.165945493Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288}
{"msg":"lookup srv","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:17:36.165964013Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"component":"DnsResolver","dns_name":"_clickhouse._tcp.control-plane.oxide.internal"}
{"msg":"failed to create ClickHouse client","v":0,"name":"oximeter","level":40,"time":"2023-10-31T19:17:36.24154484Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"file":"oximeter/collector/src/lib.rs:775","error":"Database(DatabaseVersionMismatch { expected: 3, found: 2 })","retry_after":"368.429321ms"}
{"msg":"creating ClickHouse client","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:17:36.611656623Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288}
{"msg":"lookup srv","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:17:36.611706913Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"component":"DnsResolver","dns_name":"_clickhouse._tcp.control-plane.oxide.internal"}
{"msg":"failed to create ClickHouse client","v":0,"name":"oximeter","level":40,"time":"2023-10-31T19:17:36.65835238Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"file":"oximeter/collector/src/lib.rs:775","error":"Database(DatabaseVersionMismatch { expected: 3, found: 2 })","retry_after":"657.142555ms"}
{"msg":"creating ClickHouse client","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:17:37.316457226Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288}
```

The DB is still at version 2, but the new `oximeter` expects version 3. It'll spin here in an infinite retry loop, using the internal services retry policy.

### Upgrading the schema

Next I tested the actual schema upgrade tool. In the `oximeter` zone:

```bash
root@oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606:~# /opt/oxide/oximeter/bin/clickhouse-schema-updater --host [fd00:1122:3344:101::e]:8123 --schema-directory /opt/oxide/oximeter/schema ls
Latest version: 2
Available versions:
 2 (reported by database)
 3 (expected by oximeter)
```

Then I applied the update:

```bash
root@oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606:~# /opt/oxide/oximeter/bin/clickhouse-schema-updater --host [fd00:1122:3344:101::e]:8123 --schema-directory /opt/oxide/oximeter/schema --log-level debug up 3
Oct 31 19:20:41.189 DEBG starting upgrade to desired version 3, id: 06ba6496-b954-4670-93e4-775bbf85f553, component: clickhouse-client, unit: clickhouse_schema_updater
Oct 31 19:20:41.189 DEBG reading entries from schema dir, dir: /opt/oxide/oximeter/schema/single-node, id: 06ba6496-b954-4670-93e4-775bbf85f553, component: clickhouse-client, unit: clickhouse_schema_updater
Oct 31 19:20:41.189 DEBG skipping non-directory, name: db-wipe.sql, id: 06ba6496-b954-4670-93e4-775bbf85f553, component: clickhouse-client, unit: clickhouse_schema_updater
Oct 31 19:20:41.189 DEBG skipping non-directory, name: db-init.sql, id: 06ba6496-b954-4670-93e4-775bbf85f553, component: clickhouse-client, unit: clickhouse_schema_updater
Oct 31 19:20:41.190 DEBG valid version dir, ver: 3, id: 06ba6496-b954-4670-93e4-775bbf85f553, component: clickhouse-client, unit: clickhouse_schema_updater
Oct 31 19:20:41.190 DEBG valid version dir, ver: 2, id: 06ba6496-b954-4670-93e4-775bbf85f553, component: clickhouse-client, unit: clickhouse_schema_updater
Oct 31 19:20:41.190 DEBG reading SQL files from schema dir, dir: /opt/oxide/oximeter/schema/single-node/3, id: 06ba6496-b954-4670-93e4-775bbf85f553, component: clickhouse-client, unit: clickhouse_schema_updater
Oct 31 19:20:41.195 DEBG apply schema upgrade file, filename: up, path: /opt/oxide/oximeter/schema/single-node/3/up.sql, version: 3, id: 06ba6496-b954-4670-93e4-775bbf85f553, component: clickhouse-client, unit: clickhouse_schema_updater
Oct 31 19:20:41.211 DEBG successfully applied schema upgrade file, name: up, path: /opt/oxide/oximeter/schema/single-node/3/up.sql, version: 3, id: 06ba6496-b954-4670-93e4-775bbf85f553, component: clickhouse-client, unit: clickhouse_schema_updater
Upgrade to oximeter database version 3 complete
root@oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606:~# /opt/oxide/oximeter/bin/clickhouse-schema-updater --host [fd00:1122:3344:101::e]:8123 --schema-directory /opt/oxide/oximeter/schema ls
Latest version: 3
Available versions:
 2
 3 (reported by database) (expected by oximeter)
```

And over in the `oximeter` zone, we see the service eventually detect that the database version matches the compiled version, and continue operating as normal after a restart:

```bash
{"msg":"failed to create ClickHouse client","v":0,"name":"oximeter","level":40,"time":"2023-10-31T19:18:59.682117682Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"file":"oximeter/collector/src/lib.rs:775","error":"Database(DatabaseVersionMismatch { expected: 3, found: 2 })","retry_after":"72.772067321s"}
{"msg":"creating ClickHouse client","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:20:12.456604915Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288}
{"msg":"lookup srv","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:20:12.456668875Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"component":"DnsResolver","dns_name":"_clickhouse._tcp.control-plane.oxide.internal"}
{"msg":"failed to create ClickHouse client","v":0,"name":"oximeter","level":40,"time":"2023-10-31T19:20:12.502534124Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"file":"oximeter/collector/src/lib.rs:775","error":"Database(DatabaseVersionMismatch { expected: 3, found: 2 })","retry_after":"214.749169268s"}

{"msg":"creating ClickHouse client","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.256377372Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288}
{"msg":"lookup srv","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.256941857Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"component":"DnsResolver","dns_name":"_clickhouse._tcp.control-plane.oxide.internal"}
{"msg":"registered endpoint","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.293174086Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"local_addr":"[fd00:1122:3344:101::d]:12223","component":"dropshot","path":"/info","method":"GET"}
{"msg":"registered endpoint","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.293202856Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"local_addr":"[fd00:1122:3344:101::d]:12223","component":"dropshot","path":"/producers","method":"GET"}
{"msg":"registered endpoint","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.293214806Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"local_addr":"[fd00:1122:3344:101::d]:12223","component":"dropshot","path":"/producers","method":"POST"}
{"msg":"registered endpoint","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.293224947Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"local_addr":"[fd00:1122:3344:101::d]:12223","component":"dropshot","path":"/producers/{producer_id}","method":"DELETE"}
{"msg":"listening","v":0,"name":"oximeter","level":30,"time":"2023-10-31T19:23:47.293234887Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"local_addr":"[fd00:1122:3344:101::d]:12223","component":"dropshot","file":"/home/bnaecker/.cargo/git/checkouts/dropshot-a4a923d29dccc492/fa728d0/dropshot/src/server.rs:195"}
{"msg":"successfully registered DTrace USDT probes","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.293243447Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"local_addr":"[fd00:1122:3344:101::d]:12223","component":"dropshot"}
{"msg":"contacting nexus","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.318548253Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288}
{"msg":"lookup_ipv6 srv","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.318566804Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"component":"DnsResolver","dns_name":"_nexus._tcp.control-plane.oxide.internal"}
{"msg":"accepted connection","v":0,"name":"oximeter","level":30,"time":"2023-10-31T19:23:47.366801445Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"local_addr":"[fd00:1122:3344:101::d]:12223","component":"dropshot","file":"/home/bnaecker/.cargo/git/checkouts/dropshot-a4a923d29dccc492/fa728d0/dropshot/src/server.rs:769","remote_addr":"[fd00:1122:3344:101::b]:49263"}
{"msg":"registered new metric producer","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.372961992Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","address":"[fd00:1122:3344:101::2]:12224","producer_id":"02582ccd-7613-4ca5-81e7-f88b1cc3b76b"}
{"msg":"request completed","v":0,"name":"oximeter","level":30,"time":"2023-10-31T19:23:47.373039013Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"uri":"/producers","method":"POST","req_id":"7430e7a6-dd98-4877-8934-41126ec74229","remote_addr":"[fd00:1122:3344:101::b]:49263","local_addr":"[fd00:1122:3344:101::d]:12223","component":"dropshot","file":"/home/bnaecker/.cargo/git/checkouts/dropshot-a4a923d29dccc492/fa728d0/dropshot/src/server.rs:853","latency_us":2853,"response_code":"204"}
{"msg":"registered new metric producer","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.373260895Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","address":"[fd00:1122:3344:101::c]:12221","producer_id":"52028f5e-8f19-461a-8624-1b1450f26654"}
{"msg":"request completed","v":0,"name":"oximeter","level":30,"time":"2023-10-31T19:23:47.373290286Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"uri":"/producers","method":"POST","req_id":"6f8d78cb-1d49-4ef8-90a3-1261ee8af4f2","remote_addr":"[fd00:1122:3344:101::b]:49263","local_addr":"[fd00:1122:3344:101::d]:12223","component":"dropshot","file":"/home/bnaecker/.cargo/git/checkouts/dropshot-a4a923d29dccc492/fa728d0/dropshot/src/server.rs:853","latency_us":61,"response_code":"204"}
{"msg":"registered new metric producer","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.373484947Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","address":"[fd00:1122:3344:101::a]:12221","producer_id":"a5869a7a-0e53-485e-9834-71f1b0223c07"}
{"msg":"request completed","v":0,"name":"oximeter","level":30,"time":"2023-10-31T19:23:47.373509848Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"uri":"/producers","method":"POST","req_id":"c1ce20e1-11ec-4541-9b1f-37229001213d","remote_addr":"[fd00:1122:3344:101::b]:49263","local_addr":"[fd00:1122:3344:101::d]:12223","component":"dropshot","file":"/home/bnaecker/.cargo/git/checkouts/dropshot-a4a923d29dccc492/fa728d0/dropshot/src/server.rs:853","latency_us":67,"response_code":"204"}
{"msg":"registered new metric producer","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.37376873Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","address":"[fd00:1122:3344:101::b]:12221","producer_id":"d49aee67-71d0-467d-bbcd-037c2c3d8219"}
{"msg":"request completed","v":0,"name":"oximeter","level":30,"time":"2023-10-31T19:23:47.37381235Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"uri":"/producers","method":"POST","req_id":"56be562d-514f-46ce-8e8a-7c21b702e838","remote_addr":"[fd00:1122:3344:101::b]:49263","local_addr":"[fd00:1122:3344:101::d]:12223","component":"dropshot","file":"/home/bnaecker/.cargo/git/checkouts/dropshot-a4a923d29dccc492/fa728d0/dropshot/src/server.rs:853","latency_us":131,"response_code":"204"}
{"msg":"oximeter registered with nexus","v":0,"name":"oximeter","level":30,"time":"2023-10-31T19:23:47.374598988Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"file":"oximeter/collector/src/lib.rs:847","id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606"}
{"msg":"starting oximeter collection task","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.450307176Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"producer_id":"02582ccd-7613-4ca5-81e7-f88b1cc3b76b","component":"collection-task","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","interval":"10s"}
{"msg":"starting oximeter collection task","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.451321376Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"producer_id":"a5869a7a-0e53-485e-9834-71f1b0223c07","component":"collection-task","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","interval":"10s"}
{"msg":"starting oximeter collection task","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.455914459Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"producer_id":"d49aee67-71d0-467d-bbcd-037c2c3d8219","component":"collection-task","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","interval":"10s"}
{"msg":"starting oximeter collection task","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:47.456929018Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"producer_id":"52028f5e-8f19-461a-8624-1b1450f26654","component":"collection-task","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","interval":"10s"}
{"msg":"collecting from producer","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:57.450271514Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"producer_id":"02582ccd-7613-4ca5-81e7-f88b1cc3b76b","component":"collection-task","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent"}
{"msg":"collecting from producer","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:57.450543787Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"producer_id":"a5869a7a-0e53-485e-9834-71f1b0223c07","component":"collection-task","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent"}
{"msg":"collected 1 total results","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:57.453288512Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"producer_id":"02582ccd-7613-4ca5-81e7-f88b1cc3b76b","component":"collection-task","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent"}
{"msg":"collecting from producer","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:57.456045988Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"producer_id":"d49aee67-71d0-467d-bbcd-037c2c3d8219","component":"collection-task","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent"}
{"msg":"collected 3 total results","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:57.456101749Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"producer_id":"a5869a7a-0e53-485e-9834-71f1b0223c07","component":"collection-task","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent"}
{"msg":"collecting from producer","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:57.456612023Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"producer_id":"52028f5e-8f19-461a-8624-1b1450f26654","component":"collection-task","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent"}
{"msg":"collected 3 total results","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:57.457072068Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"producer_id":"d49aee67-71d0-467d-bbcd-037c2c3d8219","component":"collection-task","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent"}
{"msg":"collected 3 total results","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:23:57.457468241Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"producer_id":"52028f5e-8f19-461a-8624-1b1450f26654","component":"collection-task","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent"}
{"msg":"inserting 131 samples into database","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:24:02.287703157Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"component":"results-sink","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent"}
{"msg":"unrolling 131 total samples","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:24:02.287768058Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"id":"ab78e276-a1ea-4002-bfaf-9674f0f00e14","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent"}
{"msg":"retrieving timeseries schema from database","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:24:02.302708728Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"id":"ab78e276-a1ea-4002-bfaf-9674f0f00e14","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent"}
{"msg":"inserted rows into table","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:24:02.331015343Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"id":"ab78e276-a1ea-4002-bfaf-9674f0f00e14","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","table_name":"oximeter.fields_i64","n_rows":40}
{"msg":"inserted rows into table","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:24:02.333399395Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"id":"ab78e276-a1ea-4002-bfaf-9674f0f00e14","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","table_name":"oximeter.fields_string","n_rows":198}
{"msg":"inserted rows into table","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:24:02.335948469Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"id":"ab78e276-a1ea-4002-bfaf-9674f0f00e14","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","table_name":"oximeter.fields_uuid","n_rows":325}
{"msg":"inserted rows into table","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:24:02.337388872Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"id":"ab78e276-a1ea-4002-bfaf-9674f0f00e14","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","table_name":"oximeter.measurements_bool","n_rows":6}
{"msg":"inserted rows into table","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:24:02.338724085Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"id":"ab78e276-a1ea-4002-bfaf-9674f0f00e14","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","table_name":"oximeter.measurements_cumulativei64","n_rows":70}
{"msg":"inserted rows into table","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:24:02.340514711Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"id":"ab78e276-a1ea-4002-bfaf-9674f0f00e14","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","table_name":"oximeter.measurements_histogramf64","n_rows":34}
{"msg":"inserted rows into table","v":0,"name":"oximeter","level":20,"time":"2023-10-31T19:24:02.341768803Z","hostname":"oxz_oximeter_f3a5ade4-a165-4ecd-a365-e57b3cdf3606","pid":22288,"id":"ab78e276-a1ea-4002-bfaf-9674f0f00e14","component":"clickhouse-client","collector_id":"f3a5ade4-a165-4ecd-a365-e57b3cdf3606","component":"oximeter-agent","table_name":"oximeter.measurements_i64","n_rows":21}
```

There are unit tests to verify this, but we can indeed see that the schema has been changed to fix #4369:

```bash
oxz_clickhouse_de7eb711-5b35-4b64-894c-7d3a4764c189 :) describe table timeseries_schema;

DESCRIBE TABLE timeseries_schema

Query id: cad647e1-4831-49a8-b63b-c99b15145905

┌─name────────────┬─type───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│ timeseries_name │ String                                                                                                                                                                                                                                                     │              │                    │         │                  │                │
│ fields.name     │ Array(String)                                                                                                                                                                                                                                              │              │                    │         │                  │                │
│ fields.type     │ Array(Enum8('Bool' = 1, 'I64' = 2, 'IpAddr' = 3, 'String' = 4, 'Uuid' = 6, 'I8' = 7, 'U8' = 8, 'I16' = 9, 'U16' = 10, 'I32' = 11, 'U32' = 12, 'U64' = 13))                                                                                                 │              │                    │         │                  │                │
│ fields.source   │ Array(Enum8('Target' = 1, 'Metric' = 2))                                                                                                                                                                                                                   │              │                    │         │                  │                │
│ datum_type      │ Enum8('Bool' = 1, 'I64' = 2, 'F64' = 3, 'String' = 4, 'Bytes' = 5, 'CumulativeI64' = 6, 'CumulativeF64' = 7, 'HistogramI64' = 8, 'HistogramF64' = 9, 'I8' = 10, 'U8' = 11, 'I16' = 12, 'U16' = 13, 'I32' = 14, 'U32' = 15, 'U64' = 16, 'F32' = 17, 'CumulativeU64' = 18, 'CumulativeF32' = 19, 'HistogramI8' = 20, 'HistogramU8' = 21, 'HistogramI16' = 22, 'HistogramU16' = 23, 'HistogramI32' = 24, 'HistogramU32' = 25, 'HistogramU64' = 26, 'HistogramF32' = 27) │              │                    │         │                  │                │
│ created         │ DateTime64(9, 'UTC')                                                                                                                                                                                                                                       │              │                    │         │                  │                │
└─────────────────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘

6 rows in set. Elapsed: 0.001 sec.

oxz_clickhouse_de7eb711-5b35-4b64-894c-7d3a4764c189 :)
```

karencfv commented on oximeter/db/schema/replicated/db-init.sql:
```diff
@@ -0,0 +1,716 @@
+    datum UInt64
+)
+ENGINE = Distributed('oximeter_cluster', 'oximeter', 'measurements_u64_local', xxHash64(splitByChar(':', timeseries_name)[1]));
+
```
We seem to be missing a few tables here like `oximeter.measurements_f32`

bnaecker replied:
Yeah, tests caught that too, thanks. Will add.

karencfv commented on oximeter/db/schema/replicated/db-wipe.sql:
This file seems to be missing a change that I recently made to add `SYNC` at the end of the line https://github.com/oxidecomputer/omicron/pull/4372/files#diff-5a50af54d5a68272a9417fa3c9eb037045c78ef7d7fda59e7fb5108e6953e0a0R1

bnaecker replied:
I think you may have been looking at an older version of the PR. `SYNC` is indeed there, as of [2063d47](https://github.com/oxidecomputer/omicron/pull/4394/commits/2063d4732b55ed49a8009c79f8fd8bda3f2687dd). See the raw file [here](https://github.com/oxidecomputer/omicron/blob/2063d4732b55ed49a8009c79f8fd8bda3f2687dd/oximeter/db/schema/replicated/db-wipe.sql)

karencfv commented on oximeter/collector/src/bin/clickhouse-schema-updater.rs:
I'm not sure I follow how this binary relates to the collector specifically, perhaps we can move it to the db directory?

bnaecker replied:
It's about responsibility over the database. `oximeter` is the main owner / writer, everyone else reads. I think we'll also want to make sure oximeter isn't running while we do these upgrades, and it's convenient that these belong in the same zone.

karencfv replied:
yep! I meant more in the sense that it's inside the `oximeter/collector/` directory and am not sure how schema upgrades fit specifically with the collector. I was wondering if it made more sense to have in the `oximeter/db/` directory instead

bnaecker replied:
It originally was, but it's slightly annoying to pull a binary from one package into the zone image for another. I think if we add more tools specifically for managing Clickhouse, then it might make sense to layer those together, but it doesn't seem critical for the one binary we have now.

karencfv commented on oximeter/db/src/client.rs:
```diff
@@ -267,14 +271,301 @@ impl Client {
+        version: u64,
+        schema_dir: impl AsRef<Path>,
+    ) -> Result<BTreeMap<String, (PathBuf, String)>, Error> {
+        let version_schema_dir = schema_dir
+            .as_ref()
+            .join(if replicated { "replicated" } else { "single-node" })
+            .join(version.to_string());
```
Let's make a helper function out of this, since it's being repeated in a function above

karencfv commented on oximeter/db/src/client.rs:
```diff
@@ -267,14 +271,301 @@ impl Client {
+        if !available.contains(&desired_version) {
+            return Err(Error::MissingSchemaVersion(desired_version));
+        }
+
```
Do we have a check making sure all available versions are consecutive numbers?

bnaecker replied:
No, we don't. What bad outcome would that protect us from, given that we only upgrade to versions strictly greater than the current?

karencfv replied:
Perhaps this is a nit, but would like to avoid a situation where someone skipped a version by mistake (1, 2, 5...)

bnaecker replied:
Sure, done in fbd898397

karencfv commented on oximeter/db/src/client.rs:
```diff
@@ -770,37 +1093,39 @@ impl DbWrite for Client {
         // The HTTP client doesn't support multiple statements per query, so we break them out here
         // manually.
```
Looks like we don't need this comment anymore

karencfv commented on oximeter/db/src/client.rs:
```diff
@@ -770,37 +1093,39 @@ impl DbWrite for Client {
 
     /// Initialize a single node telemetry database, creating tables as needed.
     async fn init_single_node_db(&self) -> Result<(), Error> {
         // The HTTP client doesn't support multiple statements per query, so we break them out here
         // manually.
```
Same as above

karencfv commented on oximeter/db/src/client.rs:
```diff
@@ -3338,4 +3665,385 @@ mod tests {
+        let schema_dir = TempDir::new().expect("failed to create tempdir");
+        let mut paths = Vec::with_capacity(versions.len());
+        for version in versions.iter() {
+            let version_dir = schema_dir
+                .path()
+                .join(if replicated { "replicated" } else { "single-node" })
+                .join(version.to_string());
```
Same about helper function

karencfv commented on oximeter/db/src/client.rs:
```diff
@@ -3338,4 +3665,385 @@ mod tests {
+        (schema_dir, paths)
+    }
+
+    #[tokio::test]
```
All of these tests currently only run against a single node installation. Now that all tests have been ported over to run against a replicated installation, these should be modified as well

bnaecker replied:
Done in e7213592a

karencfv reviewed (commented):
This is great! Really looking forward to having this feature. 

Has this been tested manually on a replicated set up as well? 

In the past we talked about generating the SQL files dynamically instead of having multiple hardcoded files, which will also aid in testing (https://github.com/oxidecomputer/omicron/issues/3982). With this approach, we seem to be relying more on hardcoded SQL files. How do you see us moving forward with both schema updates and dynamic SQL?

bnaecker commented on package-manifest.toml:
```diff
@@ -107,9 +107,12 @@ setup_hint = """
 only_for_targets.image = "standard"
 source.type = "local"
-source.rust.binary_names = ["oximeter"]
+source.rust.binary_names = ["oximeter", "clickhouse-schema-updater"]
```
It's not really a service, just a single, so creating a zone to run just that one tool seems overkill to me. This is analogous to the `nexus` and `schema-updater` binaries, both of which go into the `nexus` zone. Nexus is the main owner of the control plane database. Oximeter is the main owner of the timeseries database.

karencfv replied:
Ok, that makes sense. I was thinking of how dnswait is its own service

bnaecker commented on oximeter/collector/src/bin/clickhouse-schema-updater.rs:
```diff
@@ -0,0 +1,126 @@
+    #[clap(visible_aliases = ["up", "apply"])]
+    Upgrade {
+        /// The version to which to upgrade.
+        #[arg(default_value_t = OXIMETER_VERSION)]
+        version: u64,
```
I can see myself wanting to take it slowly on some early rollouts of this, yes.

bnaecker commented:
The replicated tests in this PR are annoyingly flaky. I'm betting there is some port conflict with other tests, since we still hardcode all port numbers in the replicated cluster setup code. I'm not able to reproduce the failure locally though, and can't really verify that since we do not currently store the tempfiles the ClickHouse processes write their logs to on failure. I'll try to repro or insert some code in the tests to dump those files to the test log we _do save, but I think the real way to resolve this is to dynamically create the cluster with any free port numbers.

karencfv reviewed (commented):
I think I know what's going on here, the serial_test crate is not compatible with nextest. See https://github.com/oxidecomputer/omicron/pull/4149#discussion_r1338101147. My best guess is that some of your ports are clashing.

I'm not sure it's a good idea to have some of the tests running against a single cluster and other tests spinning up their own clusters; we should settle on one format or the other. 

I'm concerned that by changing the format and spinning up so many clusters serially, the testing times will go through the roof.

That said, if the testing times aren't affected, then sure why not. But then, we should change all tests to avoid inconsistency.

Allocating ports dynamically sounds good, but heads up that there are many complications with this approach. Configuration files for each node are not all the same, there are many ports to allocate, and I had to reallocate some ports due to the fact that we are running all nodes in a single server.

bnaecker commented:
Thanks @karencfv that's a helpful link! I'll see if I can get test groups working for this PR.

I would really like to move everything over to running against their own cluster, the way we do for single-node tests. Looking at the tests themselves, much of the time is spent waiting for the DB to start; find the other cluster members; and possible distribute the SQL. I don't know for sure, but I don't see an obvious reason why that can't all be parallelized.

> Allocating ports dynamically sounds good, but heads up that there are many complications with this approach. Configuration files for each node are not all the same, there are many ports to allocate, and I had to reallocate some ports due to the fact that we are running all nodes in a single server.

Yeah, I'm sure it will be tricky and frustrating to get this working. That said, there is definitely support for [dynamic reconfiguration](https://clickhouse.com/docs/en/guides/sre/keeper/clickhouse-keeper#reconfiguration) of a cluster, so I think it's technically possible to have each node learn of the IPs and ports of the others.

karencfv commented:
> Thanks @karencfv that's a helpful link! I'll see if I can get test groups working for this PR.
>
> I would really like to move everything over to running against their own cluster, the way we do for single-node tests. Looking at the tests themselves, much of the time is spent waiting for the DB to start; find the other cluster members; and possible distribute the SQL. I don't know for sure, but I don't see an obvious reason why that can't all be parallelized.

Can we move this work to a separate PR? This change in testing format feels out of scope for offline schema updates. What do you think?

bnaecker commented:
I'm not planning on doing that all now, no, only the tests I have here.

bnaecker commented:
@karencfv I believe I've addressed your comment, but please let me know if I've missed something. @smklein I'd love your input as well, though I know you're still getting back into the swing of things. Thanks to you both!

karencfv commented on test-utils/src/dev/clickhouse.rs:
```diff
@@ -23,6 +23,9 @@ use crate::dev::poll;
 const CLICKHOUSE_TIMEOUT: Duration = Duration::from_secs(30);
 
+// Timeout used when starting a ClickHouse keeper subprocess.
+const CLICKHOUSE_KEEPER_TIMEOUT: Duration = Duration::from_secs(60);
```
Have you been observing longer times for the keeper to start up? That feels a bit odd, I generally observed that they spun up quicker than the servers.

bnaecker replied:
No I think this was during my attempts to understand the failures in CI. I can revert it.

bnaecker replied:
Done in 6b3d86622

karencfv commented on oximeter/db/src/client.rs:
```diff
@@ -3338,4 +3680,524 @@ mod tests {
+    async fn test_read_schema_upgrade_sql_files() {
+        let logctx = test_setup_log("test_read_schema_upgrade_sql_files");
+        let log = &logctx.log;
+        const REPLICATED: bool = false;
```
Let's test for replicated as well

bnaecker replied:
This test doesn't actually require a database, it's just using this to find the right schema subdirectory.

karencfv replied:
Yes, we want to make sure it also finds the replicated schema subdirectory as well.

karencfv commented on oximeter/db/src/client.rs:
```diff
@@ -3338,4 +3680,524 @@ mod tests {
+        let address = cluster.replica_1.address;
+        test_apply_one_schema_upgrade_impl(log, address, true).await;
+
+        // TODO-cleanup: These should be arrays.
```
If this TODO is meant for a follow up PR, can it be linked to the corresponding issue?

bnaecker replied:
Done in 6b3d86622

karencfv commented on oximeter/db/src/client.rs:
```diff
@@ -3338,4 +3680,524 @@ mod tests {
+        let logctx =
+            test_setup_log("test_ensure_schema_with_version_gaps_fails");
+        let log = &logctx.log;
+        let mut db = ClickHouseInstance::new_single_node(0)
```
Eventually we'll remove single node mode. For tests that don't require a separate test for single-node/replicated, we may want to favour replicated over single node. When the time comes to remove single node functionality, we'll have less code to remove/adapt.

bnaecker replied:
IMO, we're still a bit of time from removing single-node entirely, and I prefer having the ability to run tests individually and in parallel in the meantime.

karencfv replied:
I think I might have not explained myself very well :smile: . What I meant was that for tests like this one where it doesn't matter what type of installation it runs against, it may be a good idea to run it against a cluster instead. Basically just swapping `ClickHouseInstance::new_single_node(0)` for `ClickHouseCLuster::new()`.

Anyway, this is probably just a nit. Either way is probably fine

karencfv commented on oximeter/db/src/client.rs:
```diff
@@ -3338,4 +3680,524 @@ mod tests {
+            "test_ensure_schema_with_missing_desired_schema_version_fails",
+        );
+        let log = &logctx.log;
+        let mut db = ClickHouseInstance::new_single_node(0)
```
as above

karencfv commented on oximeter/db/src/client.rs:
```diff
@@ -3338,4 +3680,524 @@ mod tests {
+        let logctx = test_setup_log("test_select_all_field_types");
+        let log = &logctx.log;
+
+        let mut db = ClickHouseInstance::new_single_node(0)
```
as above

karencfv reviewed (commented):
Thanks for all the work with the tests @bnaecker ! I know testing replicated mode can be pretty frustrating.

Left a few comments and just want to bump a question I had in one of my previous comments:

> In the past we talked about generating the SQL files dynamically instead of having multiple hardcoded files, which will also aid in testing (https://github.com/oxidecomputer/omicron/issues/3982). With this approach, we seem to be relying more on hardcoded SQL files. How do you see us moving forward with both schema updates and dynamic SQL?

bnaecker commented:
> In the past we talked about generating the SQL files dynamically instead of having multiple hardcoded files, which will also aid in testing (https://github.com/oxidecomputer/omicron/issues/3982). With this approach, we seem to be relying more on hardcoded SQL files. How do you see us moving forward with both schema updates and dynamic SQL?

I've not thought too much about it, but it doesn't seem to preclude generating the SQL. For example, one could imagine always generating the _last_ version of the SQL from a build.rs script that uses the current supported measurement and field types.

karencfv commented:
> I've not thought too much about it, but it doesn't seem to preclude generating the SQL. For example, one could imagine always generating the last version of the SQL from a build.rs script that uses the current supported measurement and field types.

Do you see us also keeping the SQL upgrade files in the directories as well?

bnaecker commented:
Probably? I don't know how one would generate _upgrade_ SQL statements programmatically, or if it's worth trying to do so. What I'm trying to prevent is the issues we've seen like missing tables; missing enum variants; distributed tables pointing at the wrong local table; etc. Those seem to apply to the last SQL more than any upgrade statements.

bnaecker commented:
Gentle ping on this one @smklein, I'd love your feedback before merging!

smklein commented on oximeter/db/src/client.rs:
```diff
@@ -267,14 +271,320 @@ impl Client {
+        )
+        .await?;
+
+        // We need to be pretty careful at this point with any data-modifying
```
What would be wrong with inserting a placeholder row? I basically did this to insert the bootstrap "version" of the database

bnaecker replied:
I'm not sure I understand your question. Are you asking why we're validating anything about the queries? Or specifically about inserting a row in the version table?

smklein replied:
Kinda, yeah, like, what would be wrong with a data-modifying statement to "insert one row"?

bnaecker replied:
I think the short answer is that it's not clear to me one could do that and still have a valid set of data in the tables.

The `oximeter` collector normalizes the samples it receives, and always inserts records into at least 2 tables for any sample (at least one field, and the measurement). I think it's important for `oximeter` to be responsible for that, to ensure that its invariants are upheld. The version table is managed by the upgrade process, so I'm not sure we need to allow that now. The schema table is also managed automatically by `oximeter`, as it derives the schema from each sample.

I've been thinking about how to manage things like adding new columns that do require generating data. In those cases, I think we can relax this to adding something like a default expression to the `ADD COLUMN` statement. It's also possible that we want to _move_ data around, which would involve something like `INSERT INTO`. We can relax that when we need to.

smklein commented on oximeter/collector/src/lib.rs:
```diff
@@ -454,9 +453,39 @@ impl OximeterAgent {
+        // - The database exists and is at the expected version. Continue in
+        // this case.
+        //
+        // - The database exists and is at a lower-than-expected version. We
+        // fail back to the caller here, which will retry indefinitely until the
+        // DB has been updated.
```
Nice, so until we have support for "online upgrade", this prevents Clickhouse from ever running with an old schema, right?

bnaecker replied:
That's right. Oximeter will spin on startup until the DB version matches its compiled version.

smklein reviewed (approved):
Apologies for the delay, looks good to me!

bnaecker commented on oximeter/db/schema/README.md:
````diff
@@ -0,0 +1,40 @@
+ oxControlService20/omicron6 fd00:1122:3344:101::e/64
+ ```
+- Log into the `oximeter` zone, `zlogin oxz_oximeter_<UUID>`
+- Ensure `oximeter` is _not_ running, e.g., `svcadm disable oximeter`
````
It could, yes, and I think that's OK. (Unless I'm missing something!) The updater tool accepts an address, and so as long as the oximeter zone can reach that, it should work fine. I'm actually not sure that this is important anymore though. I was trying to be quite conservative, but I think with the additional check in `oximeter`'s startup process, which waits for the DB to arrive at the current version, this warning may no longer be relevant. WDYT?

smklein replied:
Yeah, I saw the logic later that stopped Oximeter from fully booting until the version matched, so it's probably okay to just omit these instructions?

bnaecker replied:
Done in fa280fd73

bnaecker commented on oximeter/db/src/client.rs:
```diff
@@ -267,14 +271,320 @@ impl Client {
+    fn verify_schema_upgrades(
+        files: &BTreeMap<String, (PathBuf, String)>,
+    ) -> Result<(), Error> {
+        let re =
+            regex::Regex::new("(INSERT INTO)|(ALTER TABLE .* DELETE)").unwrap();
```
To the first question, yes, I can make it `static`.

For the latter question, I was again being pretty conservative. The point was to, for now, disallow any data modifications in these updates themselves. If that is a goal we agree on, then yeah, I should probably add lightweight deletes and other mutations to the data itself. But perhaps that's just not a worthwhile guardrail now?

bnaecker replied:
Made things static and added more checks for mutations in fa280fd73
